crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -393,7 +393,9 @@ impl IndexerReader {
     }
 
     /// Retrieve the system state data for the given epoch. If no epoch is given,
-    /// it will retrieve the last known epoch's data and return the system state.
+    /// it will retrieve the latest epoch's data and return the system state.
+    /// System state of the current epoch is populated at the beginning of the epoch,
```

Comment:
Sorry @gegaowp, but I'm still confused.  Is the system_state row for epoch E populated at the beginning of epoch E, or epoch E+1?

If it's populated at the beginning epoch E+1, then it seems like we should not offer the option to pass `None` to this API, because it will always error.

If it's populated at epoch E, then we have a problem because it will appear as if the system state for epoch E "jumps back in time" when E is no longer the latest system state -- any pending state that accumulated over the course of the epoch will be lost because they will get applied for the new latest system state and will not be included in the historical state.

Reply:
@amnn the system state of epoch E is populated at beginning of epoch E indeed

> any pending state that accumulated over the course of the epoch will be lost

system state at end of epoch E is the same as at the beginning of epoch E + 1, I can change it to override the beginning state at the end of epoch so that for historic epoch, we will read the "end of epoch" state, and for current, we will read the beginning one, would that work? The only small caveat is that, we can no longer read system state at genesis.

Reply:
Got it, now I understand.  There's a difference between what you have suggested here:

> system state at end of epoch E is the same as at the beginning of epoch E + 1

And what I really want, which is to remember the system state at the end of epoch E, just before the beginning of epoch E + 1.  Is it possible to do that? I think with that definition of historical system state, there wouldn't be such an issue with the genesis state (which I agree it's important to preserve) because the historical system state for the first epoch will include the parameters from genesis, and then some incremental changes that happened in the first epoch, which would be easy to distinguish from each other.

Reply:
> which is to remember the system state at the end of epoch E, just before the beginning of epoch E + 1

this is what happens today, when I assign "system state at the beginning of epoch E + 1", it actually reading system state from the last checkpoint data of epoch E (which is in memory rather than committed data in DB) https://github.com/MystenLabs/sui/blob/e24ee790999a04645d9d526cb164a6818e2cbc38/crates/sui-indexer/src/handlers/checkpoint_handler_v2.rs#L235

Reply:
Okay great, in that case, this:

> I can change it to override the beginning state at the end of epoch so that for historic epoch, we will read the "end of epoch" state, and for current, we will read the beginning one, would that work?

Sounds great -- or I would even go further and say that you don't even need to write to the `system_state` field at the beginning of the epoch, only at the end -- this would help GraphQL identify when the row of the epoch table it is reading corresponds to the latest epoch.

I also want to make sure then that this:

> I think with that definition of historical system state, there wouldn't be such an issue with the genesis state (which I agree it's important to preserve) because the historical system state for the first epoch will include the parameters from genesis, and then some incremental changes that happened in the first epoch, which would be easy to distinguish from each other.

Matches your understanding as well -- or whether this approach really does risk messing up the initial system state stored irrecoverably?

Reply:
@amnn sounds good, updated the PR and it only writes at the end of epochs.

> whether this approach really does risk messing up the initial system state stored irrecoverably?

sounds reasonable assuming all changed parameters within epoch 0 was just from 0 to a non-zero number, while value non-zero values from genesis indeed persists thru the epoch.

Reply:
Great!  In that case, I think this function's comment needs to be tweaked, and its `epoch` parameter should be mandatory (because querying the latest one will now fail), but the other changes on the PR look good!